### PR TITLE
Decoding Structs

### PR DESCRIPTION
**Pull request by mmsc2**
Decodification of operands

### CODE REVIEW

File: src/instruction.c [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#include "instruction.h"
+
+unsigned int instruction_size(Instruction ins) {
+	if (ins.op1_addr == Imm)
+		return 2;
+	else
+		return 1;
+}
```
Comment (jrchatruc) [inline src/instruction.c]: This function is not being used, let's remove it (and with it the `instruction.c` file)

File: test/test.c [hunk 2]
```diff
@@ -1,10 +1,248 @@
+#include "decoder.h"
 #include <assert.h>
 #include <stdio.h>
 
-int main(int argc, char **argv)
-{
-	assert(1);
-	printf("Tests passed!\n");
+void decoding_tests(void) {
```
Comment (jrchatruc) [inline test/test.c]: Move this function to a separate file, called something like `instruction_decoding_test.c`

File: test/test.c [hunk 3]
```diff
@@ -1,10 +1,248 @@
...
+
+	printf("--------------------------------- \n");
+	printf("Test: non_zero_high_bit \n");
+
+	ResultInstruction err = decode_instruction(0x94A7800080008000);
+	VirtualMachineError vm_error = err.value.error;
+	if (err.is_error)
+		assert(1);
+	if (vm_error == InstructionNonZeroHighBit)
+		assert(1);
```
Comment (jrchatruc) [inline test/test.c]: This won't make the test fail if the behaviour is wrong, instead just call
```
assert (err.is_error)
assert(vm_error == InstructionNonZeroHighBit
```

and so on for the rest of the tests

File: src/instruction.c (hunk 1 above)
Comment (mmsc2) [reply src/instruction.c]: done

File: test/test.c (hunk 2 above)
Comment (mmsc2) [reply test/test.c]: done

File: test/test.c (hunk 3 above)
Comment (mmsc2) [reply test/test.c]: done

File: src/instruction.h [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+typedef struct Instruction Instruction;
+typedef struct ResultInstruction ResultInstruction;
+typedef enum VirtualMachineError VirtualMachineError;
+typedef union ResultInstructionValue ResultInstructionValue;
+
+// ------------------
+// 	  Functions
+// ------------------
+
+unsigned int instruction_size(Instruction ins);
```
Comment (jrchatruc) [inline src/instruction.h]: We removed this function, remove the declaration as well

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**